env/context.go

```diff
@@ -5,59 +5,174 @@
 package env
 
 import (
+	"fmt"
 	"net"
+	"os"
+	"path/filepath"
 	"sync"
 
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/go-framed-msgpack-rpc/rpc"
 )
 
-// Context is an implementation for libkbfs.Context
-type Context struct {
+const (
+	kbfsSocketFile = "kbfsd.sock"
+)
+
+// Context defines the environment for this package
+type Context interface {
+	GetRunMode() libkb.RunMode
+	GetLogDir() string
+	GetDataDir() string
+	ConfigureSocketInfo() (err error)
+	GetSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error)
+	NewRPCLogFactory() *libkb.RPCLogFactory
+	// TODO: uncomment once we've added changes to allow gobind to create a
+	// KBFS context from a *libkb.GlobalContext
+	//GetKBFSSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error)
+	//BindToKBFSSocket() (net.Listener, error)
+}
+
+// KBFSContext is an implementation for libkbfs.Context
+type KBFSContext struct {
 	g *libkb.GlobalContext
+	// protects the socket primitives below
+	kbfsSocketMtx     sync.RWMutex
+	kbfsSocket        libkb.Socket
+	kbfsSocketWrapper *libkb.SocketWrapper
+}
+
+var _ Context = (*KBFSContext)(nil)
+
+func (c *KBFSContext) initKBFSSocket() {
+	c.kbfsSocketMtx.Lock()
+	defer c.kbfsSocketMtx.Unlock()
+	log := c.g.Log
+	bindFile := c.getkbfsSocketFile()
+	dialFiles := []string{bindFile}
+	c.kbfsSocket = libkb.NewSocketWithFiles(log, bindFile, dialFiles)
 }
 
-var libkbOnce sync.Once
+// NewContextFromGlobalContext constructs a context
+func NewContextFromGlobalContext(g *GlobalContext) *KBFSContext {
+	g.Init()
+	g.ConfigureConfig()
+	g.ConfigureLogging()
+	g.ConfigureCaches()
+	g.ConfigureMerkleClient()
```

Comment:
What breaks if you don't run these routines?

Reply:
I could spend some time digging into it, but if someone already knows that would be easier.

Reply:
Did you try at first without this, and then fail?

Reply:
oh i see it down there, ok...

Reply:
why not just run `g.Init(); g.Configure()` from main?

Reply:
I'll propose what I think is a much simpler change, one sec...